File: creational/abstract-factory/abstract-factory.java
public class AbstractFactoryPattern {
    public interface Breed {
        String getBreed();
    }

    public static class Siamese implements Breed {
        public String getBreed() {
            return "Siamese";
        }
    }

    public static class MaineCoon implements Breed {
        public String getBreed() {
            return "Maine Coon";
        }
    }

    public static class Alabai implements Breed {
        public String getBreed() {
            return "Alabai";
        }
    }

    public static class Retriever implements Breed {
        public String getBreed() {
            return "Retriever";
        }
    }

    public interface Origin {
        String getOrigin();
    }

    public static class Thailand implements Origin {
        public String getOrigin() {
            return "Thailand";
        }
    }

    public static class Maine implements Origin {
        public String getOrigin() {
            return "Maine, United States of America";
        }
    }

    public static class Kazakhstan implements Origin {
        public String getOrigin() {
            return "Kazakhstan";
        }
    }

    public static class Canada implements Origin {
        public String getOrigin() {
            return "Canada";
        }
    }

    public interface QueryComponentsFactory {
        Breed setCatBreed();
        Origin setCatOrigin();
        Breed setDogBreed();
        Origin setDogOrigin();
    }

    public static class AsianComponentsFactory implements QueryComponentsFactory {
        public Breed setCatBreed() {
            return new Siamese();
        }

        public Origin setCatOrigin() {
            return new Thailand();
        }

        public Breed setDogBreed() {
            return new Alabai();
        }

        public Origin setDogOrigin() {
            return new Kazakhstan();
        }
    }

    public static class AmericanComponentsFactory implements QueryComponentsFactory {
        public Breed setCatBreed() {
            return new MaineCoon();
        }

        public Origin setCatOrigin() {
            return new Maine();
        }

        public Breed setDogBreed() {
            return new Retriever();
        }

        public Origin setDogOrigin() {
            return new Canada();
        }
    }

    public static abstract class AnimalQuery {
        protected String species;
        protected Breed breed;
        protected Origin origin;
        protected QueryComponentsFactory componentsFactory;

        public AnimalQuery(QueryComponentsFactory componentsFactory) {
            this.componentsFactory = componentsFactory;
        }

        public abstract void composeQuery();

        public void showDescription() {
            String description = String.format("%s is a %s breed from %s.", breed.getBreed(), species, origin.getOrigin());
            System.out.println(description);
        }
    }

    public static class CatQuery extends AnimalQuery {
        public CatQuery(QueryComponentsFactory componentsFactory) {
            super(componentsFactory);
            species = "cat";
        }

        public void composeQuery() {
            breed = componentsFactory.setCatBreed();
            origin = componentsFactory.setCatOrigin();
        }
    }

    public static class DogQuery extends AnimalQuery {
        public DogQuery(QueryComponentsFactory componentsFactory) {
            super(componentsFactory);
            species = "dog";
        }

        public void composeQuery() {
            breed = componentsFactory.setDogBreed();
            origin = componentsFactory.setDogOrigin();
        }
    }

    public static void fulfillQuery(AnimalQuery query) {
        query.composeQuery();
        query.showDescription();
    }

    public static void main(String[] args) {
        QueryComponentsFactory americanComponentsFactory = new AmericanComponentsFactory();
        QueryComponentsFactory asianComponentsFactory = new AsianComponentsFactory();

        AnimalQuery siameseQuery = new CatQuery(asianComponentsFactory);
        fulfillQuery(siameseQuery);

        AnimalQuery maineCoonQuery = new CatQuery(americanComponentsFactory);
        fulfillQuery(maineCoonQuery);

        AnimalQuery alabaiQuery = new DogQuery(asianComponentsFactory);
        fulfillQuery(alabaiQuery);

        AnimalQuery retrieverQuery = new DogQuery(americanComponentsFactory);
        fulfillQuery(retrieverQuery);
    }
}

File: creational/abstract-factory/abstract-factory.ts
interface Breed {
    getBreed(): string;
}

class Siamese implements Breed {
    public getBreed(): string {
        return 'Siamese';
    }
}

class MaineCoon implements Breed {
    public getBreed(): string {
        return 'Maine Coon';
    }
}

class Alabai implements Breed {
    public getBreed(): string {
        return 'Alabai';
    }
}

class Retriever implements Breed {
    public getBreed(): string {
        return 'Retriever';
    }
}

interface Origin {
    getOrigin(): string;
}

class Thailand implements Origin {
    public getOrigin(): string {
        return 'Thailand';
    }
}

class Maine implements Origin {
    public getOrigin(): string {
        return 'Maine, United States of America';
    }
}

class Kazakhstan implements Origin {
    public getOrigin(): string {
        return 'Kazakhstan';
    }
}

class Canada implements Origin {
    public getOrigin(): string {
        return 'Canada';
    }
}

interface QueryComponentsFactory {
    setCatBreed(): Breed;
    setCatOrigin(): Origin;
    setDogBreed(): Breed;
    setDogOrigin(): Origin;
}

class AsianComponentsFactory implements QueryComponentsFactory {
    public setCatBreed(): Breed {
        return new Siamese();
    }

    public setCatOrigin(): Origin {
        return new Thailand();
    }

    public setDogBreed(): Breed {
        return new Alabai();
    }

    public setDogOrigin(): Origin {
        return new Kazakhstan();
    }
}

class AmericanComponentsFactory implements QueryComponentsFactory {
    public setCatBreed(): Breed {
        return new MaineCoon();
    }

    public setCatOrigin(): Origin {
        return new Maine();
    }

    public setDogBreed(): Breed {
        return new Retriever();
    }

    public setDogOrigin(): Origin {
        return new Canada();
    }
}

abstract class AnimalQuery {
    protected abstract _species: string;
    protected _breed: Breed;
    protected _origin: Origin;
    protected _componentsFactory: QueryComponentsFactory;

    constructor(componentsFactory: QueryComponentsFactory) {
        this._componentsFactory = componentsFactory;
    }

    public abstract composeQuery(): void;

    public showDescription(): void {
        console.log(`${this._breed.getBreed()} is a ${this._species} breed from ${this._origin.getOrigin()}.`);
    }
}

class CatQuery extends AnimalQuery {
    protected _species: string = 'cat';

    public composeQuery(): void {
        this._breed = this._componentsFactory.setCatBreed();
        this._origin = this._componentsFactory.setCatOrigin();
    }
}

class DogQuery extends AnimalQuery {
    protected _species: string = 'dog';

    public composeQuery(): void {
        this._breed = this._componentsFactory.setDogBreed();
        this._origin = this._componentsFactory.setDogOrigin();
    }
}

let americanComponentsFactory: QueryComponentsFactory = new AmericanComponentsFactory();
let asianComponentsFactory: QueryComponentsFactory = new AsianComponentsFactory();

const fulfillQuery: (query: AnimalQuery) => void = (query: AnimalQuery) => {
    query.composeQuery();
    query.showDescription();
}

let siameseQuery: AnimalQuery = new CatQuery(asianComponentsFactory);
fulfillQuery(siameseQuery);

let maineCoonQuery: AnimalQuery = new CatQuery(americanComponentsFactory);
fulfillQuery(maineCoonQuery);

let alabaiQuery: AnimalQuery = new DogQuery(asianComponentsFactory);
fulfillQuery(alabaiQuery);

let retrieverQuery: AnimalQuery = new DogQuery(americanComponentsFactory);
fulfillQuery(retrieverQuery);
